In [1]:
from IPython.core.debugger import set_trace

%load_ext nb_black

import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import time
import keras

plt.style.use(style="seaborn")
%matplotlib inline

<IPython.core.display.Javascript object>

In [3]:
train = pd.read_csv(f"tok.csv")


<IPython.core.display.Javascript object>

In [4]:
train.head().T

0  \
target                                                  0   
txt     awthats a bummer you shoulda got david carr of...   

                                                        1  \
target                                                  0   
txt     is upset that he cant update his facebook by t...   

                                                        2  \
target                                                  0   
txt     i dived many times for the ball managed to sav...   

                                                      3  \
target                                                0   
txt     my whole body feels itchy and like its on fire    

                                                        4  
target                                                  0  
txt     its not behaving at all im mad why am i here b...

<IPython.core.display.Javascript object>

### Basic NLP

In [6]:
from collections import Counter

# Count unique words
def counter_word(text):
    count = Counter()
    for i in text.values:
        for word in i.split():
            count[word] += 1
    return count

<IPython.core.display.Javascript object>

In [7]:
text = train.txt

counter = counter_word(text)

<IPython.core.display.Javascript object>

In [8]:
len(counter)

555350

<IPython.core.display.Javascript object>

In [9]:
15000 / len(counter)

0.027009993697668138

<IPython.core.display.Javascript object>

In [10]:
for item in counter.most_common(15000):
    print(item)

('i', 694479)
('to', 546705)
('the', 499559)
('a', 364346)
('my', 301258)
('and', 268321)
('you', 251188)
('is', 228093)
('it', 210973)
('for', 208140)
('in', 205969)
('of', 176812)
('im', 163293)
('on', 160881)
('me', 149480)
('have', 138461)
('so', 137604)
('that', 132669)
('just', 117990)
('with', 110558)
('be', 107662)
('at', 107366)
('was', 100024)
('not', 98190)
('its', 97551)
('but', 94992)
('this', 88338)
('up', 83151)
('good', 83065)
('now', 79807)
('day', 79023)
('get', 78274)
('out', 76269)
('all', 75311)
('are', 74541)
('like', 73676)
('go', 69903)
('no', 64796)
('dont', 63124)
('do', 62424)
('your', 62334)
('going', 61694)
('love', 59957)
('cant', 59176)
('got', 58256)
('today', 57974)
('work', 57857)
('too', 57597)
('from', 54355)
('am', 53565)
('back', 52888)
('time', 52001)
('u', 50342)
('lol', 49860)
('what', 48801)
('one', 48661)
('will', 48603)
('about', 48435)
('we', 47785)
('know', 47046)
('really', 46562)
('had', 45775)
('can', 44460)
('see', 43584)
('some', 43464

('keys', 532)
('character', 532)
('shooting', 531)
('texts', 531)
('impossible', 531)
('bugs', 531)
('justin', 530)
('loose', 530)
('swift', 530)
('fa', 530)
('insomnia', 529)
('fo', 529)
('fully', 529)
('steve', 529)
('shorts', 528)
('attempt', 528)
('que', 527)
('picnic', 527)
('theyve', 527)
('cruise', 527)
('lab', 527)
('stream', 527)
('lessons', 527)
('gaga', 527)
('dan', 526)
('bio', 525)
('hm', 525)
('adding', 525)
('admit', 525)
('fill', 524)
('cinema', 524)
('arrive', 524)
('cleaned', 524)
('pleased', 524)
('graduate', 524)
('sittin', 523)
('blow', 523)
('mcdonalds', 523)
('hits', 522)
('saving', 521)
('vids', 521)
('response', 521)
('inspiration', 521)
('yoga', 521)
('nail', 520)
('uploaded', 520)
('crush', 519)
('couldve', 519)
('someday', 519)
('himself', 519)
('tweetie', 519)
('drank', 519)
('bell', 519)
('besides', 519)
('kni', 518)
('skool', 518)
('marketing', 518)
('idiot', 517)
('seats', 517)
('umm', 516)
('infection', 516)
('rofl', 516)
('cupcakes', 516)
('gd', 515)
(

('naps', 227)
('icky', 227)
('bullshit', 227)
('edinburgh', 227)
('pi', 227)
('concept', 227)
('ro', 227)
('stalking', 227)
('grilled', 227)
('sp', 227)
('lens', 227)
('cranky', 227)
('it´s', 227)
('severe', 226)
('psp', 226)
('bump', 226)
('fry', 226)
('ink', 226)
('pepper', 226)
('noi', 226)
('looove', 226)
('tiger', 226)
('choc', 226)
('sub', 226)
('legit', 226)
('horses', 225)
('inch', 225)
('hyper', 225)
('ignored', 225)
('eek', 225)
('hou', 225)
('rap', 225)
('cracking', 225)
('particular', 225)
('logged', 225)
('zac', 225)
('wierd', 225)
('bat', 225)
('cuts', 225)
('rb', 225)
('gardening', 224)
('overall', 224)
('concerned', 224)
('chicks', 224)
('ay', 224)
('balance', 224)
('greet', 224)
('device', 224)
('brad', 224)
('daisies', 224)
('shave', 223)
('dev', 223)
('tools', 223)
('postponed', 223)
('lush', 223)
('nut', 223)
('ross', 223)
('soap', 223)
('european', 223)
('loli', 223)
('transfer', 223)
('lambert', 223)
('powers', 222)
('owl', 222)
('punk', 222)
('hn', 222)
('shaking

('catchy', 123)
('cider', 123)
('leftover', 123)
('mileys', 123)
('mw', 123)
('icarly', 123)
('yethe', 123)
('inspirational', 122)
('underneath', 122)
('fallon', 122)
('facing', 122)
('conditioning', 122)
('entries', 122)
('hamilton', 122)
('spirits', 122)
('hotels', 122)
('podcasts', 122)
('juicy', 122)
('launched', 122)
('wifes', 122)
('hsm', 122)
('dearly', 122)
('bittersweet', 122)
('wrestling', 122)
('bruises', 122)
('ohhhhh', 122)
('cheeky', 122)
('raid', 122)
('bleed', 122)
('columbus', 122)
('streak', 122)
('crown', 122)
('firmware', 122)
('diving', 122)
('penis', 122)
('howdy', 122)
('whyyy', 122)
('influence', 122)
('soz', 122)
('confuse', 122)
('mov', 122)
('gloom', 122)
('watches', 122)
('flatley', 122)
('tourney', 121)
('noah', 121)
('pas', 121)
('ce', 121)
('anyones', 121)
('unwell', 121)
('mau', 121)
('banks', 121)
('migraines', 121)
('dozen', 121)
('eeek', 121)
('sock', 121)
('stranded', 121)
('tila', 121)
('sao', 121)
('spoilers', 121)
('cotton', 121)
('doom', 121)
('h

('bargain', 79)
('honda', 79)
('orchestra', 79)
('atcha', 79)
('destiny', 79)
('loveyou', 79)
('garros', 79)
('heavenly', 79)
('neda', 79)
('wwwmeasia', 79)
('shareholder', 79)
('anoop', 78)
('moneys', 78)
('ulcer', 78)
('believed', 78)
('eliminated', 78)
('owen', 78)
('commit', 78)
('tyson', 78)
('evolution', 78)
('looooong', 78)
('ankles', 78)
('decorating', 78)
('stir', 78)
('superstar', 78)
('bord', 78)
('trousers', 78)
('disappearing', 78)
('vicky', 78)
('slei', 78)
('harlem', 78)
('elevator', 78)
('logan', 78)
('yayyyyy', 78)
('showtime', 78)
('booboo', 78)
('cologne', 78)
('creation', 78)
('vagina', 78)
('lawyer', 78)
('ia', 78)
('discovering', 78)
('lk', 78)
('specifically', 78)
('mattress', 78)
('haiku', 78)
('aired', 78)
('stud', 78)
('protected', 78)
('quotthats', 78)
('offi', 78)
('hiphop', 78)
('rascal', 78)
('immense', 78)
('vera', 78)
('expenses', 78)
('manic', 78)
('supportive', 78)
('alexs', 78)
('pressed', 78)
('landlord', 78)
('clumsy', 78)
('mustard', 78)
('mornibut

('promoted', 57)
('katies', 57)
('craziest', 57)
('actin', 57)
('whatev', 57)
('attic', 57)
('brandi', 57)
('amandas', 57)
('psh', 57)
('whod', 57)
('berkeley', 57)
('villa', 57)
('abandon', 57)
('elderly', 57)
('browns', 57)
('noted', 57)
('armed', 57)
('towed', 57)
('matey', 57)
('bobut', 57)
('thoug', 57)
('aggressive', 57)
('daybut', 57)
('moaning', 57)
('bourbon', 57)
('sparkling', 57)
('gays', 57)
('raving', 57)
('wander', 57)
('sil', 57)
('technique', 57)
('weekei', 57)
('ties', 57)
('youbut', 57)
('diaper', 57)
('umy', 57)
('bootleg', 57)
('baaad', 57)
('bobs', 57)
('yesi', 57)
('thks', 57)
('grounds', 57)
('randomness', 57)
('foxtel', 57)
('stumbled', 57)
('commence', 57)
('legacy', 57)
('kylie', 57)
('goodi', 57)
('producers', 57)
('mute', 57)
('shouted', 57)
('quothes', 57)
('strips', 57)
('stunt', 57)
('quotgo', 57)
('enemies', 57)
('jax', 57)
('subtle', 57)
('scrapbook', 57)
('addresses', 57)
('eternity', 57)
('giggling', 57)
('macys', 57)
('swineflu', 57)
('loops', 57)
('

('oceans', 42)
('solving', 42)
('chee', 42)
('kellan', 42)
('winters', 42)
('assist', 42)
('blossom', 42)
('bleak', 42)
('bottled', 42)
('slowest', 42)
('tomrrow', 42)
('richie', 42)
('arrrgh', 42)
('nav', 42)
('koolaid', 42)
('tyres', 42)
('postage', 42)
('lexington', 42)
('knocks', 42)
('chains', 42)
('sbux', 42)
('stuart', 42)
('bragging', 42)
('franz', 42)
('olives', 42)
('grannys', 42)
('mick', 42)
('fridbut', 42)
('midterms', 42)
('fink', 42)
('encouraged', 42)
('transferred', 42)
('packets', 42)
('norris', 42)
('resistance', 42)
('altogether', 42)
('tou', 42)
('muchhh', 42)
('secondary', 42)
('cured', 42)
('mana', 42)
('stomachs', 42)
('bldg', 42)
('albany', 42)
('notch', 42)
('shawty', 42)
('rumble', 42)
('cox', 42)
('nhs', 42)
('bev', 42)
('aber', 42)
('halp', 42)
('marmite', 42)
('slightest', 42)
('sephora', 42)
('gak', 42)
('strait', 42)
('chico', 42)
('summa', 42)
('sly', 42)
('thoughi', 42)
('goits', 42)
('dino', 42)
('morningg', 42)
('insects', 42)
('exploding', 42)
('res

<IPython.core.display.Javascript object>

In [11]:
num_words = 15000

# Max number of words in a sequence
max_length = 20

<IPython.core.display.Javascript object>

In [12]:
train_size = int(train.shape[0] * 0.8)

train_sentences = train.txt[:train_size]
train_labels = train.target[:train_size]

test_sentences = train.txt[train_size:]
test_labels = train.target[train_size:]

<IPython.core.display.Javascript object>

In [13]:
from keras.preprocessing.text import Tokenizer

tokenizer = Tokenizer(num_words=num_words)
tokenizer.fit_on_texts(train_sentences)

<IPython.core.display.Javascript object>

In [14]:
word_index = tokenizer.word_index

<IPython.core.display.Javascript object>

In [16]:
train_sequences = tokenizer.texts_to_sequences(train_sentences)

<IPython.core.display.Javascript object>

In [17]:
from keras.preprocessing.sequence import pad_sequences

train_padded = pad_sequences(
    train_sequences, maxlen=max_length, padding="post", truncating="post"
)

<IPython.core.display.Javascript object>

In [18]:
test_sequences = tokenizer.texts_to_sequences(test_sentences)
test_padded = pad_sequences(
    test_sequences, maxlen=max_length, padding="post", truncating="post"
)

<IPython.core.display.Javascript object>

In [20]:
early_stopping = keras.callbacks.EarlyStopping(
    monitor="val_loss", patience=5, restore_best_weights=True, verbose=0
)

# checkpoint_callback = keras.callbacks.ModelCheckpoint(
#         filepath='.', monitor='val_loss', verbose=0, save_weights_only=True,
#         save_freq='epoch', mode='auto', save_best_only=True)

reduce_lr_on_plateau = keras.callbacks.ReduceLROnPlateau(
    monitor="acc", factor=0.1, patience=2, verbose=0
)

callbacks_list = [early_stopping, reduce_lr_on_plateau]

<IPython.core.display.Javascript object>

In [22]:
import keras_tuner as kt
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import (
    Flatten,
    Dense,
    Embedding,
    Conv1D,
    GlobalMaxPooling1D,
    MaxPooling1D,
)


def build_model(hp):

    model = Sequential()
    model.add(
        Embedding(
            num_words,
            output_dim=hp.Int("output_dim", min_value=128, max_value=1024, step=128),
            input_length=max_length,
        )
    )
    model.add(
        Conv1D(
            filters=hp.Int("filters0", min_value=64, max_value=512, step=64),
            kernel_size=hp.Int("kernel_size0", min_value=1, max_value=3, step=1),
            padding="same",
            activation="relu",
            strides=hp.Int("strides0", min_value=1, max_value=4, step=1),
        )
    )
    model.add(MaxPooling1D(pool_size=2, padding='same'))
    if hp.Boolean("conv1"):
        model.add(
            Conv1D(
                filters=hp.Int("filters1", min_value=32, max_value=256, step=32),
                kernel_size=hp.Int("kernel_size1", min_value=1, max_value=3, step=1),
                padding="same",
                activation="relu",
                strides=hp.Int("strides1", min_value=1, max_value=4, step=1),
            )
        )
        model.add(MaxPooling1D(pool_size=2, padding='same'))
    model.add(Flatten())
    model.add(
        Dense(
            units=hp.Int("units0", min_value=128, max_value=512, step=64),
            activation="relu",
        )
    )
#     if hp.Boolean("dense1"):
    model.add(
        Dense(
            units=hp.Int("units1", min_value=64, max_value=512, step=64),
            activation="relu",
        )
    )
    if hp.Boolean("dense2"):
        model.add(
            Dense(
                units=hp.Int("units2", min_value=64, max_value=256, step=32),
                activation="relu",
            )
        )    
    model.add(Dense(1, activation="sigmoid"))
    model.compile(optimizer="rmsprop", loss="binary_crossentropy", metrics=["acc"])
    return model
########################################units????


# model.add(GlobalMaxPooling1D())
# model.compile(optimizer="rmsprop", loss="binary_crossentropy", metrics=["acc"])
# model.summary()
    build_model(kt.HyperParameters())


<IPython.core.display.Javascript object>

In [25]:
tuner = kt.Hyperband(
    build_model,
    "val_loss",
    30,
    factor=3,
    hyperband_iterations=1,
)

INFO:tensorflow:Reloading Oracle from existing project .\untitled_project\oracle.json
INFO:tensorflow:Reloading Tuner from .\untitled_project\tuner0.json


<IPython.core.display.Javascript object>

In [26]:
tuner.search(train_padded, train_labels, validation_data=(test_padded, test_labels), callbacks=callbacks_list, verbose=1)


Trial 7 Complete [01h 48m 57s]
val_loss: 0.7181563377380371

Best val_loss So Far: 0.49407514929771423
Total elapsed time: 07h 34m 49s

Search: Running Trial #8

Hyperparameter    |Value             |Best Value So Far 
output_dim        |128               |768               
filters0          |256               |256               
kernel_size0      |2                 |3                 
strides0          |3                 |3                 
conv1             |False             |True              
units0            |384               |192               
units1            |256               |128               
dense2            |True              |False             
filters1          |128               |32                
kernel_size1      |2                 |1                 
strides1          |2                 |1                 
units2            |128               |None              
tuner/epochs      |2                 |2                 
tuner/initial_e...|0                 |0 

KeyboardInterrupt: 

<IPython.core.display.Javascript object>

In [ ]:
# history = model.fit(
#     train_padded,
#     train_labels,
#     epochs=20,
#     validation_data=(test_padded, test_labels),
# )

Trial 7 Complete [01h 48m 57s]
val_loss: 0.7181563377380371

Best val_loss So Far: 0.49407514929771423
Total elapsed time: 07h 34m 49s

Search: Running Trial #8

Hyperparameter    |Value             |Best Value So Far 
output_dim        |128               |768               
filters0          |256               |256               
kernel_size0      |2                 |3                 
strides0          |3                 |3                 
conv1             |False             |True              
units0            |384               |192               
units1            |256               |128               
dense2            |True              |False             
filters1          |128               |32                
kernel_size1      |2                 |1                 
strides1          |2                 |1                 
units2            |128               |None              
tuner/epochs      |2                 |2                 
tuner/initial_e...|0                 |0                 
tuner/bracket     |3                 |3                 
tuner/round       |0                 |0                 

Epoch 1/2
 5591/39910 [===>..........................] - ETA: 10:17 - loss: 0.5234 - acc: 0.7445